### DISCUSSION TITLE
problem configure charon vs charon-nm

### DISCUSSION THREAD

**Discussion by plabanowski** [strongswan/strongswan#801] [Q&A]
Hello,

I have a problem using charon-nm. I would like to configure the connection with networkmanager to establish an EAP-GTC connection.

it works in command line using strongswan directly:

- it is the conf of my connection ipsec.conf:
```
conn vpn1
        type=tunnel
        auto=add
        leftfirewall=yes
        forceencaps=yes
        leftauth=eap-gtc
        leftid=*****
        left=%any
        leftsourceip=%config
        rightauth=pubkey
        right=****
        rightsubnet=****
        keyexchange=ikev2
        eap_identity=****
        ike=aes256-sha256-modp2048
        esp=aes256-sha256
        ikelifetime=24h
        lifetime=1h
        reauth=yes
        rekey=yes
        margintime=1m
        rekeyfuzz=50%
        dpdaction=restart
        dpddelay=30s
        dpdtimeout=60s
```
- and here is the log of a successful connection with strongwan:
```
parsed IKE_AUTH response 1 [ IDr CERT AUTH EAP/REQ/ID ]
**** certif *** crl fetching failed
authentication of '**********' with RSA signature successful
server requested EAP_IDENTITY (id 0x72), sending '**********'
generating IKE_AUTH request 2 [ EAP/RES/ID ]
sending packet: from ********** (80 bytes)
received packet: from ********** (80 bytes)
parsed IKE_AUTH response 2 [ EAP/REQ/GTC ]
server requested EAP_GTC authentication (id 0x73)
generating IKE_AUTH request 3 [ EAP/RES/GTC ]
sending packet: from ********** (96 bytes)
received packet: from ********** (80 bytes)
parsed IKE_AUTH response 3 [ EAP/SUCC ]
EAP method EAP_GTC succeeded, no MSK established
authentication of '**********' (myself) with EAP
generating IKE_AUTH request 4 [ AUTH ]
sending packet: from 1**********[4500] (112 bytes)
received packet: from **********(800 bytes)
parsed IKE_AUTH response 4 [ AUTH N(CRASH_DET) CPRP(ADDR DNS DNS) SA TSi TSr N(ESP_TFC_PAD_N) N(NON_FIRST_FRAG) ]
authentication of '**********' with EAP successful
IKE_SA vpn1[3] established between **********
scheduling reauthentication in 86336s
maximum IKE_SA lifetime 86396s
installing DNS server ********** via resolvconf
installing DNS server ********** via resolvconf
installing new virtual IP 1**********
received ESP_TFC_PADDING_NOT_SUPPORTED, not using ESPv3 TFC padding
selected proposal: ESP:AES_CBC_256/HMAC_SHA2_256_128/NO_EXT_SEQ
CHILD_SA vpn1{4} established with SPIs c89673cf_i ac1b78d3_o and TS **********
```

- Well now with NetworkManager which uses charon-nm and not charon if I understood correctly.
The log :
```
charon-nm[224166]: 05[IKE] initiating IKE_SA strongswan EAP[1] to **********
charon-nm[224166]: 05[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
charon-nm[224166]: 05[NET] sending packet: from ********** (464 bytes)
NetworkManager[223898]: <info>  [1638737582.8979] vpn-connection[0x55e647a7c110,a6690237-4978-41a8-b6af-a2c193b27c0a,"strongswan EAP",0]: VPN plugin: state changed: starting (3)
charon-nm[224166]: 10[NET] received packet: from ********** (525 bytes)
charon-nm[224166]: 10[ENC] parsed IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) CERTREQ ]
charon-nm[224166]: 10[CFG] selected proposal: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_2048
charon-nm[224166]: 10[IKE] local host is behind NAT, sending keep alives
charon-nm[224166]: 10[IKE] received 5 cert requests for an unknown ca
charon-nm[224166]: 10[IKE] establishing CHILD_SA strongswan EAP{1}
charon-nm[224166]: 10[ENC] generating IKE_AUTH request 1 [ IDi N(INIT_CONTACT) IDr N(IPCOMP_SUP) SA TSi TSr N(MOBIKE_SUP) N(ADD_4_ADDR) N(ADD_4_ADDR) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
charon-nm[224166]: 10[NET] sending packet: from ********** (368 bytes)
charon-nm[224166]: 12[IKE] retransmit 1 of request with message ID 1
charon-nm[224166]: 12[NET] sending packet: from* ********* (368 bytes)
charon-nm[224166]: 13[IKE] retransmit 2 of request with message ID 1
charon-nm[224166]: 13[NET] sending packet: from ********** (368 bytes)
charon-nm[224166]: 15[IKE] retransmit 3 of request with message ID 1
charon-nm[224166]: 15[NET] sending packet: from **********] (368 bytes)
charon-nm[224166]: 07[IKE] sending keep alive to **********[4500]
```
if you have an idea, because I do not see what block 

I don't understand this log :
`charon-nm[224166]: 10[ENC] generating IKE_AUTH request 1 [ IDi N(INIT_CONTACT) IDr N(IPCOMP_SUP) SA TSi TSr N(MOBIKE_SUP) N(ADD_4_ADDR) N(ADD_4_ADDR) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]`

and why it doesn't want to do EAP-GTC

thanks for your expertise

**Comment by tobiasbrunner** [comment]
Check the log of the responder. It either does not receive the IKE_AUTH request or there is a problem with the response (note that the server does not enable IKEv2 fragmentation, so there might be an issue there if the server certificate is large).

> and why it doesn't want to do EAP-GTC

That's requested by the server via IKE_AUTH responses (you see that in the other log).

**Comment by plabanowski** [comment]
Thank you for your answer. I tried to look at the log files on the server side of checkpoint (R81), but I find nothing to do. 
It works correctly with the strongswan client in command line, impossible to redo the configuration with networkmanager.
I have to look more in detail where are the complete vpn logs in the checkpoint box.

**Comment by madman9999** [comment]
Just checked it with R81.20 and Ubuntu 22.04

strongSwan: 5.9.5
Netzwork Manager StrongSwan Plugin: 1.52

Please add the following into /etc/strongswan.conf

**********************************
# strongswan.conf - strongSwan configuration file
#
# Refer to the strongswan.conf(5) manpage for details
#
# Configuration changes should be made in the included files


charon-nm {
	load_modular = yes
        send_vendor_id = yes
        plugins {
                include strongswan.d/charon/*.conf
                attr {
                    dns = 192.168.0.21, 192.168.0.31
                }
        }

}

charon {
  	load_modular = yes
  	send_vendor_id = yes
  	plugins {
    		include strongswan.d/charon/*.conf
		attr {
	            dns = 192.168.0.21, 192.168.0.31
        	}
  	}

}
include strongswan.d/*.conf
**********************************


* modify the DNS Server I
* charon-nm is used by the Network Manager
* charon is used for CLI (ipsec -up home) for example. Not used by the Network Manager

then: systemctl restart ipsec
